### ISSUE TITLE
Better error handling in constructors and null values

### ISSUE THREAD

**Issue by Gaff** [immutables/immutables#1261]
I'm having issues understanding error caused when deserialising complex JSON messages into immutables. The crux of the issue is null values in lists, but the details are hard to follow. Here's a reduction of the issue:

```java
    @Value.Immutable
    @JsonDeserialize(as=ImmutableWidget.class)
    interface Widget {
        List<String> strings();
    }

    @Test
    public void streamItNulls() throws IOException {
        String json = "{\"strings\":[\"one\",null,\"two\"]}";
        ObjectMapper mapper = new ObjectMapper();
        mapper.registerModule(new GuavaModule());
        Widget widget = mapper.readValue(json, Widget.class);
    }
```

In this simple test we attempt to deserialze an array that contains nulls, but the exception we get back is:

```
com.fasterxml.jackson.databind.exc.InvalidDefinitionException: Cannot construct instance of `ImmutableWidget`, problem: `java.lang.NullPointerException`
 at [Source: (String)"{"strings":["one",null,"two"]}"; line: 1, column: 30]

	at com.fasterxml.jackson.databind.exc.InvalidDefinitionException.from(InvalidDefinitionException.java:67)
        <....skip...>
Caused by: java.lang.NullPointerException
	at com.google.common.base.Preconditions.checkNotNull(Preconditions.java:878)
	at com.google.common.collect.ImmutableList$Builder.add(ImmutableList.java:786)
	at com.google.common.collect.ImmutableList$Builder.add(ImmutableList.java:748)
	at com.google.common.collect.ImmutableCollection$Builder.addAll(ImmutableCollection.java:441)
	at com.google.common.collect.ImmutableList$Builder.addAll(ImmutableList.java:833)
	at ImmutableWidget$Builder.addAllStrings(ImmutableWidget.java:247)
	at ImmutableWidget.fromJson(ImmutableWidget.java:141)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:567)
	at com.fasterxml.jackson.databind.introspect.AnnotatedMethod.call1(AnnotatedMethod.java:114)
	at com.fasterxml.jackson.databind.deser.std.StdValueInstantiator._createUsingDelegate(StdValueInstantiator.java:524)
	... 29 more
```

The message: `Cannot construct instance of ImmutableWidget, problem: `java.lang.NullPointerException` is not very helpful if you have dozens of fields!

If we swicth to `jdkOnly` we get the more helpful error: `Cannot construct instance of `com.mercuria.marketdata.curveservice.entities.ImmutableWidget`, problem: strings element` - this is because in this mode the immutables library does the null checking and provides a slither of helpful information it this fails.

Would it be desirable do you think to add some error colour around the setter methods, maybe generate code like so:

```
    public final Builder addAllStrings(Iterable<String> elements) {
      try {
        this.strings.addAll(elements);
        return this;
      } catch (RuntimeException ex) {
        throw new RuntimeException("Could not set strings element " + ex.toString(), ex);
      }
    }
```

**Comment by elucash** [comment]
I might be completely wrong here (so please correct me, etc) but it seems that it's `GuavaModule` and _Jackson_ internals working here to construct an instance, i.e. the control flow is outside of the code generated by Immutables: happens outside of any generated constructor/factory method or builder initializer.

**Comment by elucash** [comment]
```java
try {
        this.strings.addAll(elements);
        return this;
      } catch (RuntimeException ex) {
        throw new RuntimeException("Could not set strings element " + ex.toString(), ex);
      }
```
I would be quite reluctant we add such kind of code and in this particular example `strings element` reported in the other part of the code, so it's hard for me to stitch it together. I dunno, the checks in generated code should be fail-fast precondition checks with only the stack trace being the context. If we need to add more of deserialization document context, maybe it's good idea to write custom modules with enhanced reporting.
Simple & minimal solution I feel can help is to provide element index in `"strings element"` kind of exceptions.

**Comment by Gaff** [comment]
I agree it's a tricky call. You could argue the information is all in the stacktrace, but we have scenarios where we expect less technical users to spot issues with JSON documents and we find this case particularly cryptic for them to follow.

Another option would be to automagically annotate the element with `@JsonSetter(contentNulls = Nulls.FAIL)` so that jackson will fail in this scenario before it reaches the immutables constructor? We might end up doing this but it's a non-obvious manual thing that needs to be done all over the place - which is exactly the sort of thing we employ the Immutables library to avoid!

Is it easy to drop in custom modules for better reporting for this sort of thing? Can you point me to some examples?

**Comment by elucash** [comment]
`@JsonSetter(contentNulls = Nulls.FAIL)` is this is something we can generate on the synthetic `Json` instance: you can take generated code `ImmutableYxz` (with Json shim generated), then disable generatation for that `Yxz` and paste the generated source into regular source, then add some annotations like `JsonSetter` and you can see if it works. If it is we can easily add those into codegen template as a clear improvement.

As for creating custom modules. I've looked into those once or twice, but I've mostly worked with Gson type adapter factories or custom binding which are remote equivalent. The idea behind is that modules are like factories for serializers/deserializers, where you can create your own instances wrapping around the ones you want to delegate to, then in own instances for lists/maps/objects you can add a lot more context so if you have exceptions/errors popping out, the error messages can be enriched with extra tracking information `$.x.y[1].a` or at least better format the error (I sometimes even remove a lot of stack trace elements/causes if the problem is very clearly identifiable)
